src/main/java/racingcar/util/RandomNumberGenerator.java

```diff
@@ -0,0 +1,10 @@
+package racingcar.util;
+
+import java.security.SecureRandom;
+
+public class RandomNumberGenerator {
```

Comment:
- 유틸 클래스라면 private 생성자를 추가해서 인스턴스를 생성하지 못하도록 막아주면 더 좋습니다 🙂 
```suggestion
public class RandomNumberGenerator {
    private RandomNumberGenerator() {
    }
```
- nit : 유틸 클래스는 왜 class로 만들어야 할까요? enum으로 만들면 안 될까요? 굳이 private 생성자로 막아둘 필요도 없고 더 좋은 거 아닐까요?

Reply:
`유틸 클래스라면 private 생성자를 추가해서 인스턴스를 생성하지 못하도록 막아주면 더 좋습니다 🙂`
- 불필요한 인스턴스화를 막기 위해 private 생성자를 추가 하는 부분에 동의합니다!

`유틸 클래스는 왜 class로 만들어야 할까요? enum으로 만들면 안 될까요? 굳이 private 생성자로 막아둘 필요도 없고 더 좋은 거 아닐까요?`
- 우선, 제가 생각하는 `enum`에 대해 말씀드리겠습니다. `step1` 에서 빙봉의 코멘트에 대답하지 못했던 내용이 하나 있었는데, [nit : 에러 메시지를 enum으로 빼면 어떤 점이 좋은가요?](https://github.com/woowacourse/java-racingcar/pull/440/files/08f7466b8f55b67f0fdde3a44f0b0116f134d7a7#diff-d684fbe5ac86556c91bbd2979d73a67e4a8e96237325cfed2fdd8e742ff4295a) 입니다.
저는 enum을 `서로 연관된 상수들의 집합` 이라고 간주해서, "에러 메세지"라는 동일한 성격의 상수들을 모아 두었었습니다. 근데 Stirng 타입의 역할(Exception 생성시, String으로만 활용)만 하는 상수들을 `Enum`으로 빼면 어떤 점이 좋은지 고민해보며 대답을 못 했습니다. 만약, `throw new CustomException(ErrorType.NULL_INPUT)` 처럼 `enum 인스턴스`를 사용하는 예외나 객체를 만들었다면, Enum의 장점이 살아나겠지만 단순히 String 타입의 `message`만 사용해서 불필요하다고 느꼈습니다.

`서로 연관된 상수들의 집합` 으로 생각했을때, 아래와 같이 현재 유틸클래스`RandomNumberGenerator`를 enum으로 만들어 볼 수 있을 것 같습니다. (랜덤 숫자를 생성하지만, 알고리즘이 다른 숫자 생성기의 집합으로 생각했습니다)

```java
public enum RandomNumberGenerator {
    RAMDOM((a, b) -> new Random().nextInt(a - 2) + b),
    SECURE_RANDOM((a, b) -> new SecureRandom().nextInt(a - 2) + b);

    private final BiFunction<Integer, Integer, Integer> generate;

    NumberGenerator(BiFunction<Integer, Integer, Integer> generate) {
        this.generate = generate;
    }

    public int getNumberInRange(int a, int b) {
        return generate.apply(a, b);
    }
}
```
이렇게 하면 관계들을 명확하게 표현할 수 있고, `RandomNumberGenerator.getNumberInRange()` -> `RandomNumberGenerator.RAMDOM.getNumberInRange()` 이런식으로 호출할 수 있을 것 같습니다.

작성하면서 드는 생각은 굳이 필요한가 입니다. 변경사항이 적고 "상수"의 역할을 한다면 enum의 장점(리팩토링 최소화, 타입 체크, 컴파일 단계에서의 오류 발견,..)이 많이 활용될 것 같습니다.
위 케이스에서 만약 `특정 숫자 이하의 랜덤 숫자 생성`, `이상의 랜덤 숫자 생성` 등 메서드가 추가된다면 열거되어있는 타입마다 모두 정의해주어야 합니다. 가독성 부분에서도 RANDOM(methodA(), methodB(), methodC(), ...) 처럼 계속 추가되어 불편할 것 같습니다.
대답을 정리해보면서 내린 결론은, 변동성이 많고 행동이 많은 경우엔 생성자를 막아두고 class로 가져가는 것이 좋다고 생각하는데 빙봉은 어떻게 생각하시나요? 제가 빙봉의 물음에 잘못 이해한 부분이 있다면 알려주세요!

Reply:
음.. 조금 다르게 생각한 부분이 있는데요. 저는 RandomNumberGenerator를 열거의 기능이 아닌 유틸 컴포넌트 자체로 보는 것으로 생각했었습니다. 현재 코드를 본다면 단순히 SecureRandom 상수를 하나 선언하고, static 메서드를 통해 함수를 호출해서 사용하는 것 뿐이니까요.

즉, class에서 모든 메서드를 static하게 만들어서 유틸성 컴포넌트를 만든 경우와 enum으로 유틸성 컴포넌트를 만든 경우를 생각했습니다.

```java
// class에서 모든 메서드를 static하게 만든 경우
final class RandomNumberGenerator {
    private RandomNumberGenerator{}
    public static xxx() {}
}

// enum으로 만든 경우
public enum RandomNumberGenerator {
;
    public static xxx() {}
}
```

이렇게 되면 보통 생각하기에는 상속 불가, 인스턴스화 불가로 동작이 같을 것 같은데 어떤 게 더 나을까? 하는 질문이었습니다ㅎㅎ

이것에 대해 미리 답변하자면 저는 enum을 더 선호하는데요. 그 이유는 private 생성자를 만들더라도 인스턴스화가 완전히 불가하진 않아서입니다. 밖에서 private 생성자를 만들더라도 내부 정적 클래스를 만든다면 그 안에서는 인스턴스 생성이 가능하거든요. 이런 부분에서 인스턴스화가 아예 불가능한 상황이 오지 않도록 유틸성 컴포넌트들은 enum으로 간단하게 선언해서 작성하는 걸 좋아합니다.

Reply:
앗, 빙봉의 의도에 대해 오해 했었군요😂. 말씀 주셔서 감사해요
제가 너무 `enum`을 "열거형"이라는 집합의 개념에 집착했던 것 같습니다. 빙봉의 리뷰처럼 class 대신 enum으로 선언하여 객체를 상수처럼 선언할 수 있는 방법도 있었군요. 전혀 생각치 못한 부분이라 해당 리뷰가 더 감사하고 귀하게 느껴집니다.

저도 enum을 더 선호할 것 같습니다. 위 방식은 해당 메서드를 호출하는 방법은 동일하면서, 좀 더 안전하게 인스턴스화를 방지할 수 있을 것 같아요. 마치 class에 final을 붙힌 것 같군요. enum을 이렇게 활용할 수 있다니, 좋은 리뷰 감사드려요. `java enum vs enum`과 같은 주제에 대해 조금의 이해도가 생긴 것 같습니다!